### PR TITLE
BAU: Remove X509CertificateConfiguration class, merge CertificateDetail into Certificate

### PR DESCRIPTION
**Pull request by kerrr**
Remove `X509CertificateConfiguration` class. We feel this class is doing more harm than good.  It only adds headers and strips whitespace (to create a so-called full certificate), and there is only one place where that format is needed, so the formatting may as well be performed there instead; this also means we no longer need code for undoing that formatting.  Making this change should help simplify the setup of some tests we want to add soon.

The `Certificate` and `CertificateDetail` classes did not have clear roles and responsibilities, blending into each other, `CertificateDetail` composing certificate with some extra values. This has been flattened with those values merged into `Certificate`

So we now have:
- Simple Strings for base64 encoded representations of certificates in config classes.
- The `Certificate` class that gives us a richer representation of a certificate with added metadata of issuer entity id, usage, etc. for general use within the Config Service.
- Java.security `X509Certificate` as the actual certificate.

Review by commits. The first commit removes `X509CertificateConfiguration`, the second merges `Certificate` and `CertificateDetail`. The second commit is quite a big refactor, but with minimal changes that are not a result of the merge.

### CODE REVIEW

**Comment by kerrr** [comment]
Codacy is complaining about existing test method names that I'm not going to change.

File: hub/config/src/main/java/uk/gov/ida/hub/config/resources/CertificatesResource.java [hunk 1]
```diff
@@ -161,4 +161,17 @@ private void addCertificateHealthCheckDtos(
...
         }
     }
+
+    private Collection<Certificate> getCertificates(Set<TransactionConfig> transactionConfigs, Set<MatchingServiceConfig> matchingServiceConfigs) {
+        return Stream.concat(transactionConfigs.stream(), matchingServiceConfigs.stream())
+                .flatMap(this::getCertificates)
+                .collect(Collectors.toSet());
+    }
+
+    private Stream<Certificate> getCertificates(CertificateConfigurable config){
```
Comment (chrisclayson) [inline hub/config/src/main/java/uk/gov/ida/hub/config/resources/CertificatesResource.java]: Is this duplicate code from the config object?

File: hub/config/src/test/java/uk/gov/ida/hub/config/domain/builders/IdentityProviderConfigDataBuilder.java [hunk 2]
```diff
@@ -5,16 +5,13 @@
...
-import uk.gov.ida.hub.config.domain.SignatureVerificationCertificate;
 
-import java.util.ArrayList;
 import java.util.List;
 
 
 public class IdentityProviderConfigDataBuilder {
 
     private String entityId = "default-idp-entity-id";
-    private List<SignatureVerificationCertificate> signatureVerificationCertificates = new ArrayList<>();
```
Comment (alan-gds) [inline hub/config/src/test/java/uk/gov/ida/hub/config/domain/builders/IdentityProviderConfigDataBuilder.java]: It looks to me like this was redundant even before your changes - is that correct?

File: hub/config/src/main/java/uk/gov/ida/hub/config/resources/CertificatesResource.java (hunk 1 above)
Comment (kerrr) [reply hub/config/src/main/java/uk/gov/ida/hub/config/resources/CertificatesResource.java]: Yup,  forgot to clear this instance up.

File: hub/config/src/main/java/uk/gov/ida/hub/config/domain/Certificate.java [hunk 3]
```diff
@@ -1,84 +1,97 @@
...
     private static final Logger LOG = LoggerFactory.getLogger(Certificate.class);
     private static final String FINGERPRINT_ALGORITHM = "SHA-256";
 
-    protected String cert;
-
-    @ValidationMethod(message = "Certificate was not a valid x509 cert.")
-    public boolean isX509Valid() {
+    private final String issuerEntityId;
+    private final FederationEntityType federationEntityType;
+    private final String base46EncodeCertificate;
```
[the remark this answers is not in the document]
Comment (kerrr) [reply hub/config/src/main/java/uk/gov/ida/hub/config/domain/Certificate.java]: Doh, caught myself doing that twice already!

File: hub/config/src/main/java/uk/gov/ida/hub/config/domain/Certificate.java [hunk 4]
```diff
@@ -1,84 +1,97 @@
...
+        this.base46EncodeCertificate = base46EncodeCertificate;
+        this.certificateType = certificateType;
+        this.enabled = enabled;
+        X509Certificate x509 = null;
         try {
-            getCertificate();
+            x509 = getCertificate(base46EncodeCertificate);
         } catch (CertificateException e) {
-            return false;
+
```
[the remark this answers is not in the document]
Comment (kerrr) [reply hub/config/src/main/java/uk/gov/ida/hub/config/domain/Certificate.java]: Sure

[the remark this answers is not in the document]
Comment (kerrr) [reply hub/config/src/main/java/uk/gov/ida/hub/config/domain/Certificate.java]: Looking at this again, I've refactored the `getCertificate` method to not throw.

File: hub/config/src/main/java/uk/gov/ida/hub/config/domain/EncryptionCertificate.java [hunk 5]
```diff
@@ -1,16 +0,0 @@
-package uk.gov.ida.hub.config.domain;
-
-public class EncryptionCertificate extends Certificate {
```
[the remark this answers is not in the document]
Comment (kerrr) [reply hub/config/src/main/java/uk/gov/ida/hub/config/domain/EncryptionCertificate.java]: We weren't actually using that type safety though. We always accessed handled the returned values as the super class. So we just had a layer of complexity when we don't actually care.

File: hub/config/src/test/java/uk/gov/ida/hub/config/application/CertificateServiceTest.java [hunk 6]
```diff
@@ -264,7 +259,7 @@ public void signatureVerificationCertificatesForEntityIdThrowsWhenMatchingSignat
     }
 
     @Test
-    public void getAllCertificateDetailsReturnsAllTransactionAndMatchingServiceCertificateDetails() {
+    public void getAllCertificateReturnsAllTransactionAndMatchingServiceCertificate() {
```
[the remark this answers is not in the document]
Comment (kerrr) [reply hub/config/src/test/java/uk/gov/ida/hub/config/application/CertificateServiceTest.java]: Yup

File: hub/config/src/test/java/uk/gov/ida/hub/config/domain/builders/IdentityProviderConfigDataBuilder.java (hunk 2 above)
Comment (kerrr) [reply hub/config/src/test/java/uk/gov/ida/hub/config/domain/builders/IdentityProviderConfigDataBuilder.java]: Yes, it's been left over from the dark ages.

File: hub/config/src/main/java/uk/gov/ida/hub/config/domain/Certificate.java [hunk 7]
```diff
@@ -1,84 +1,97 @@
...
+                return "";
             }
-            return CertificateExpiryStatus.OK;
-        } catch (CertificateException e) {
-            return CertificateExpiryStatus.CRITICAL;
+        }
+        return "";
+    }
+
+    private X509Certificate getCertificate(String base64EncodeCertificate) throws CertificateException {
```
[the remark this answers is not in the document]
Comment (kerrr) [reply hub/config/src/main/java/uk/gov/ida/hub/config/domain/Certificate.java]: yup

**Pull request merged**